### PR TITLE
Feat/search nft

### PR DESCRIPTION
**Pull request by hydraCy**

### CODE REVIEW

File: Maskbook/Scene/Wallet/WalletSend/ViewModel/SearchSingleNFTViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        collectiblesPublisher = SearchSingleNFTViewModel.collectiblesPublisher(contractAdress: contractAddress)
+
+        collectiblesPublisher?
+            .receive(on: DispatchQueue.main)
+            .sink(receiveValue: { [weak self] collectibles in
+                self?.collectiblesSubject.value = collectibles
+            })
+            .store(in: &disposeBag)
+
+        Publishers.CombineLatest(searchString, collectiblesSubject)
```
Comment (foxsin10) [inline Maskbook/Scene/Wallet/WalletSend/ViewModel/SearchSingleNFTViewModel.swift]:
```suggestion
        Publishers.CombineLatest(searchString, collectiblesSubject)
            .map { text, collectibles -> [Collectible] in
                guard let text = text,
                      !text.isEmpty else {
                    return collectibles
                }
                
                return collectibles.filter { collectible in
                    guard let name = collectible.name,
                          let tokenID = collectible.tokenId else {
                        return false
                    }

                    if let _ = [name, tokenID].first(where: { $0.containsIgnoreCase(string: text) }) {
                        return true
                    } else {
                        return false
                    }
                }
            }
            .bind(to: \.dataSource, on: self)
            .store(in: &disposeBag)
```

Comment (foxsin10) [inline Maskbook/Scene/Wallet/WalletSend/ViewModel/SearchSingleNFTViewModel.swift]: And we should add `debounce` or `throttle` to searchString publisher

File: Maskbook/Supporting Files/Coordinator.swift [hunk 2]
```diff
@@ -885,6 +887,16 @@ extension Coordinator {
         case let .composeSelectContact(selectContactViewModel):
             return SelectContactViewController(viewModel: selectContactViewModel)
             
+        case let .luckydropSearchCollection(delegate: viewController):
+            let searchCollectionVc = SearchNFTCollectionViewController()
+            searchCollectionVc.delegate = viewController
+            return searchCollectionVc
+            
+        case let .luckydropSearchNFT(delegate: viewController,contractAddress: address):
```
Comment (foxsin10) [inline Maskbook/Supporting Files/Coordinator.swift]:
```suggestion
        case let .luckydropSearchNFT(selectionDelegate, contractAddress):
```

File: Maskbook/Supporting Files/Coordinator.swift [hunk 3]
```diff
@@ -885,6 +887,16 @@ extension Coordinator {
         case let .composeSelectContact(selectContactViewModel):
             return SelectContactViewController(viewModel: selectContactViewModel)
             
+        case let .luckydropSearchCollection(delegate: viewController):
```
Comment (foxsin10) [inline Maskbook/Supporting Files/Coordinator.swift]:
```suggestion
        case let .luckydropSearchCollection(selectionDelegate):
```

File: Maskbook/Scene/Wallet/WalletSend/SearchSingleNFTViewController.swift [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
+
+    private lazy var infoStackView = HStackView(distribution: .fillProportionally) {
+        selectAllButton
+        nftAmountLabel
+    }
+
+    lazy var searchTextField: MaskTextField = {
```
Comment (foxsin10) [inline Maskbook/Scene/Wallet/WalletSend/SearchSingleNFTViewController.swift]:
```suggestion
    private lazy var searchTextField: MaskTextField = {
```

Comment (foxsin10) [reply Maskbook/Scene/Wallet/WalletSend/SearchSingleNFTViewController.swift]: And also apply to all UI components which are not called from outside

File: Maskbook/Scene/Wallet/WalletSend/SearchNFTCollectionViewController.swift [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+class SearchNFTCollectionViewController: BaseViewController {
+    typealias Section = SearchCollectionNFTViewModel.Section
+    typealias Item = SearchCollectionNFTViewModel.CollectionListModel
+    typealias ViewModel = SearchCollectionNFTViewModel
+    var viewModel: ViewModel!
+    var subscriptions: Set<AnyCancellable> = []
+    weak var delegate: ChooseCollectionBackDelegate?
+
+    let searchTextField: MaskTextField = {
```
Comment (foxsin10) [inline Maskbook/Scene/Wallet/WalletSend/SearchNFTCollectionViewController.swift]:
```suggestion
    private let searchTextField: MaskTextField = {
```

Comment (foxsin10) [reply Maskbook/Scene/Wallet/WalletSend/SearchNFTCollectionViewController.swift]: And also apply to all UI components which are not called from outside

File: Maskbook/Scene/App/RedPacket/ViewModel/NftLuckyDropViewModel.swift [hunk 6]
```diff
@@ -362,3 +362,15 @@ extension NftLuckyDropViewModel {
...
 }
+
+extension NftLuckyDropViewModel: ChooseCollectionBackDelegate {
+    func chooseNFTCollectionAction(token: Collectible){
+        
+    }
+}
+
+extension NftLuckyDropViewModel: SearchSingleNFTDelegate {
+    func returnSelectedNFT(collectible: [Collectible]?){
```
Comment (foxsin10) [inline Maskbook/Scene/App/RedPacket/ViewModel/NftLuckyDropViewModel.swift]:
```suggestion
    func returnSelectedNFT(collectible: [Collectible]?) {
```

File: Maskbook/Scene/App/RedPacket/ViewModel/NftLuckyDropViewModel.swift [hunk 7]
```diff
@@ -362,3 +362,15 @@ extension NftLuckyDropViewModel {
         }
     }
 }
+
+extension NftLuckyDropViewModel: ChooseCollectionBackDelegate {
+    func chooseNFTCollectionAction(token: Collectible){
```
Comment (foxsin10) [inline Maskbook/Scene/App/RedPacket/ViewModel/NftLuckyDropViewModel.swift]:
```suggestion
    func chooseNFTCollectionAction(token: Collectible) {
```

**Review by foxsin10 — CHANGES_REQUESTED** [review]

File: CoreDataStack/Entity/Entity+Managed.swift [hunk 8]
```diff
@@ -302,6 +302,20 @@ extension Collectible: Managed {
         )
     }
     
+    public class func predicate(address: String, networkId: UInt64, contractAddress: String, enabled: Bool) -> NSPredicate {
```
Comment (sxiaojian88) [inline CoreDataStack/Entity/Entity+Managed.swift:305]:
```suggestion
    public class func predicate(accountAddress: String, networkId: UInt64, contractAddress: String, enabled: Bool) -> NSPredicate {
```

**Pull request merged**